### ISSUE TITLE
A false assertion should be considered as never-returning 

### ISSUE THREAD

**Issue by stephanedr** [microsoft/TypeScript#33805]
**TypeScript Version:** 3.7.0-dev.20191004

**Search Terms:** assertion "must return a value"

**Code**

```ts
function assert(condition: boolean): asserts condition {
    if (!condition) {
        throw new Error("Assertion failed");
    }
}

function test(): string {
    assert(false);
}
```

**Expected behavior:**
No error.

**Actual behavior:**
TS2355: A function whose declared type is neither 'void' nor 'any' must return a value.

**Playground Link:** https://www.typescriptlang.org/play/?ts=3.7-Beta&ssl=1&ssc=1&pln=10&pc=1#code/GYVwdgxgLglg9mABAQwM6oKYCcoAoIIAmMsCAXIgEZxwA2GyYAlBWpjqogWMaUgN4AoRCMQxgiXAEJuveM0RDRyxFAAWWOAHdEYDDoCiWTVlwAiAILpsfRMGQx6hM0wDcw0QF9B3waEi2UBioeCyIIVgwYADmih4ibDa49rSYbj6CQA

**Related Issues:**

**Comment by stephanedr** [comment]
This bug was indeed fixed up to typescript@3.7.0-dev.20191017, but has reappeared with typescript@3.7.0-dev.20191018.

**Comment by ahejlsberg** [comment]
> This bug was indeed fixed up to typescript@3.7.0-dev.20191017, but has reappeared with typescript@3.7.0-dev.20191018.

I just tried with a fresh build of the master branch and it seems to work fine.

**Comment by stephanedr** [comment]
OK, I'll check with next dev versions and come back if I still get the error.
